lib/metrics/context.js

```diff
@@ -6,43 +6,178 @@
+        })
+    }))
+  }
+
+  function getMemcached () {
+    try {
+      if (config.memcached.address === 'none') {
+        return NULL_MEMCACHED
       }
+
+      if (! _memcached) {
+        _memcached = new Memcached(config.memcached.address, {
+          timeout: 500,
+          retries: 1,
+          retry: 1000,
+          reconnect: 1000,
+          idle: config.memcached.idle,
+          namespace: 'fxa-metrics~'
+        })
+      }
+
+      return _memcached
+    } catch (err) {
+      log.error({ op: 'metricsContext.getMemcached', err: err })
+
+      return NULL_MEMCACHED
     }
+  }
+
+  /**
+   * Copies metrics context metadata to data, using either metadata
+   * passed in from a request or previously-saved metadata for a
+   * token.
+   *
+   * @param data    target object
+   * @param request request object
+   * @param event   event name
+   */
+  function copy (data, request, event) {
+    var metadata = request.payload && request.payload.metricsContext
+    var token = request.auth && request.auth.credentials
+    var doNotTrack = request.headers && request.headers.dnt === '1'
+
+    return P.resolve()
+      .then(function () {
+        if (metadata) {
+          return metadata
+        }
+
+        if (token && event) {
+          return restore(token, event)
+        }
+      })
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.copy', err: err })
+      })
+      .then(function (metadata) {
+        if (metadata) {
+          data.time = Date.now()
+          data.flow_id = metadata.flowId
+          data.flow_time = calculateFlowTime(data.time, metadata.flowBeginTime)
+          data.context = metadata.context
+          data.entrypoint = metadata.entrypoint
+          data.migration = metadata.migration
+          data.service = metadata.service
+
+          if (! doNotTrack) {
+            data.utm_campaign = metadata.utmCampaign
+            data.utm_content = metadata.utmContent
+            data.utm_medium = metadata.utmMedium
+            data.utm_source = metadata.utmSource
+            data.utm_term = metadata.utmTerm
+          }
+        }
 
-    return data
+        return data
+      })
   }
+
+  function restore (token, event) {
+    var memcached = getMemcached()
+    return memcached.getAsync(getKey(token, event))
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.restore', err: err })
+      })
+  }
+
+  /**
+   * Removes metrics context metadata for a token:event pair.
+   *
+   * @param token
+   * @param event
+   */
+  function remove (token, event) {
+    if (! token || ! event) {
+      return P.resolve()
+    }
+
+    var memcached = getMemcached()
+    return memcached.delAsync(getKey(token, event))
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.remove', err: err })
+      })
+  }
+}
+
+function getKey (token, event) {
+  return token.tokenId.toString('hex') + ':' + event
 }
```

Comment:
IIUC, we currently can't stash metrics data for the `account.verified` event because it's not associated with a token.  I wonder, if we made this accept any object with `uid` and `id` properties, like this:

```
function getKey(token, event) {
  return [token.uid, token.id, event].join(':')
}
```

Then perhaps we could also use it for the `account.verified` event by using the emailCode as the id:

```
metricsContext.save({ uid: account.uid, id: account.emailCode }, ['account.verified'], data)
```

(It's not obvious, but tokens have an `id` property that gives you `tokenId` as a hex string, for compatibility with the hawk auth plugin)